tests/test_function.py

```diff
@@ -897,6 +906,11 @@ def test_grad_3d(self):
         x = function.unravel(function.unravel(x, 0, (2, 2)), 0, (2, 1))
         self.assertEvalAlmostEqual(domain, self.grad(x, x), numpy.eye(4, 4).reshape(2, 1, 2, 2, 1, 2))
 
+    def test_grad_spaces(self):
+        X, x = mesh.line(1, space='X')
+        Y, y = mesh.line(1, space='Y')
+        self.assertEvalAlmostEqual(X * Y, self.grad(x**2 + y, x + y, spaces=iter(X.spaces)), 2 * x)
```

Comment:
Is there a reason we don't just write spaces=X.spaces? (applies to all tests below as well)

Reply:
Yes. `function.grad()` and related functions annotate the spaces argument as an iterable. I want to make sure that this is actually true. For instance, `function.laplace()` has to pass `frozenset(spaces)` to `function.grad()` and `function.div()` or the unittest will fail.